lib/game/tic_tac_toe.ex

```diff
@@ -5,38 +5,37 @@ defmodule Game.TicTacToe do
 
   alias Game.Board
   alias Game.Message
-  alias Game.Validator
 
   @welcome_message "Welcome to Tic-Tac-Toe"
 
-  @spec start(any(), any()) :: nil
-  def start(communicator, communicator_formatter, board \\ Board.setup_initial_board()) do
+  @spec start(any(), any(), struct()) :: nil
+  def start(communicator, communicator_formatter, players) do
+    board = Board.setup_initial_board()
     welcome = @welcome_message <> "\n" <> communicator_formatter.format_board(board)
     communicator.display(welcome)
 
-    play_turn(communicator, communicator_formatter, board)
+    start(communicator, communicator_formatter, board, players)
   end
 
-  @spec play_turn(any(), any(), map()) :: nil
-  defp play_turn(communicator, communicator_formatter, board) do
-    player_symbol = Board.calculate_turn(board)
-    communicator.display("Player #{player_symbol}, ")
-    player_input = String.trim(communicator.read_input())
-    case Validator.validate(board, player_input) do
+  def start(communicator, communicator_formatter, board, players) do
+    current_player = Map.get(players, Board.calculate_turn(board))
+    current_symbol = current_player.symbol
+    communicator.display("Player #{current_symbol}, ")
+    case current_player.type.valid_input(board, communicator) do
```

Comment:
I wonder if you could pattern match type and symbol like this

`%Player{symbol: current_symbol, type: player_type} = Map.get(players, Board.calculate_turn(board))`

Then your case statement would call `player_type.valid_input`